### PR TITLE
[Ruby client] add validation to nil parameter

### PR DESCRIPTION
**Pull request by linr211**
Changes of this PR:
1. Add validation to nil locale parameter in the API
2. fix issue https://github.com/vmware/singleton/issues/1479
3. fix issue https://github.com/vmware/singleton/issues/1484
4. add async to the load, this will fix issue 'doesn’t return expired message when cache is expired'
5. Change default_language to source_locale
6. Change to log level to debug when the key is missed

### CODE REVIEW

**Review by jin-jasper — DISMISSED** [review]

File: lib/sgtn-client/api/source.rb [hunk 1]
```diff
@@ -38,8 +44,11 @@ def self.getSources(component, locale)
         return items
       end
 
-      def self.loadBundles(locale)
+      def self.loadBundles(locale="en")
         SgtnClient.logger.debug "[Source][loadBundles]locale=#{locale}"
+        if locale.nil?
+          locale="en"
```
Comment (jessiejuachon) [inline lib/sgtn-client/api/source.rb:50]: Do not hardcode "en" here (or anywhere). Set and get the default from SgtnClient::Config.configurations.

File: lib/sgtn-client/api/translation.rb [hunk 2]
```diff
@@ -38,7 +44,10 @@ def self.getString_p(component, key, plural_args, locale)
         end
       end
 
-      def self.getString_f(component, key, args, locale, *optionals)
+      def self.getString_f(component, key, args, locale=SgtnClient::Config.configurations.default, *optionals)
+         if locale.nil?
```
Comment (jessiejuachon) [inline lib/sgtn-client/api/translation.rb:48]: Put this logic inside SgtnClient::Config.configurations so that you won't have to repeat it here and elsewhere.

File: lib/sgtn-client/api/translation.rb [hunk 3]
```diff
@@ -54,7 +63,10 @@ def self.getString_f(component, key, args, locale, *optionals)
          return s
       end
 
-      def self.getStrings(component, locale)
+      def self.getStrings(component, locale=SgtnClient::Config.configurations.default)
+        if locale.nil?
```
Comment (jessiejuachon) [inline lib/sgtn-client/api/translation.rb:67]: Put this logic inside SgtnClient::Config.configurations so that you won't have to repeat it here and elsewhere.

**Review by jessiejuachon — CHANGES_REQUESTED** [review]

File: lib/sgtn-client/api/source.rb [hunk 4]
```diff
@@ -38,8 +44,11 @@ def self.getSources(component, locale)
         return items
       end
 
-      def self.loadBundles(locale)
+      def self.loadBundles(locale="en")
```
[the remark this answers is not in the document]
Comment (linr211) [reply lib/sgtn-client/api/source.rb:47]: calling this method(loadBundles) is prior to loading the configuration, so has to hardcode it in the codes.

File: lib/sgtn-client/api/source.rb (hunk 1 above)
Comment (linr211) [reply lib/sgtn-client/api/source.rb:50]: calling this method(loadBundles) is prior to loading the configuration, so has to hardcode it in the codes.

File: lib/sgtn-client/api/translation.rb [hunk 5]
```diff
@@ -11,7 +11,10 @@ module Core
   
   class Translation
 
-      def self.getString(component, key, locale)
+      def self.getString(component, key, locale=SgtnClient::Config.configurations.default)
+        if locale.nil?
```
[the remark this answers is not in the document]
Comment (linr211) [reply lib/sgtn-client/api/translation.rb:15]: As the APIs like getString can be call separately, so seems not good to put the logic in SgtnClient::Config.configurations.

File: lib/sgtn-client/api/translation.rb (hunk 2 above)
Comment (linr211) [reply lib/sgtn-client/api/translation.rb:48]: As the APIs like getString can be call separately, so seems not good to put the logic in SgtnClient::Config.configurations.

File: lib/sgtn-client/api/translation.rb (hunk 3 above)
Comment (linr211) [reply lib/sgtn-client/api/translation.rb:67]: As the APIs like getString can be call separately, so seems not good to put the logic in SgtnClient::Config.configurations.

File: lib/sgtn-client/api/translation.rb [hunk 6]
```diff
@@ -23,7 +26,10 @@ def self.getString(component, key, locale)
         str
       end
 
-      def self.getString_p(component, key, plural_args, locale)
+      def self.getString_p(component, key, plural_args, locale=SgtnClient::Config.configurations.default)
+        if locale.nil?
```
[the remark this answers is not in the document]
Comment (linr211) [reply lib/sgtn-client/api/translation.rb:30]: As the APIs like getString can be call separately, so seems not good to put the logic in SgtnClient::Config.configurations.

**Review by tigershi — DISMISSED** [review]

**Review by jessiejuachon — DISMISSED** [review]

**Review by Xiaochao8 — CHANGES_REQUESTED** [review]
Please spilt this big PR into smaller ones so as to review easily.
I don't think all the issues have same priority.

File: lib/sgtn-client/api/translation.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (linr211) [reply lib/sgtn-client/api/translation.rb:15]: The API is top level because it needs to use the source locale to get source message.

**Review by tigershi — DISMISSED** [review]

File: lib/sgtn-client/api/translation.rb [hunk 7]
```diff
@@ -11,25 +11,33 @@ module Core
...
+          locale=SgtnClient::Config.configurations.default
+        end
+        locale = locale.to_s
+        SgtnClient.logger.debug "[Translation][getString]component=#{component}, key=#{key}, locale=#{locale}"
         str = getTranslation(component, key, locale)
         if str.nil?
           str = SgtnClient::Source.getSource(component, key, SgtnClient::Config.configurations.default)
           if str.nil?
-            SgtnClient.logger.error "[Translation]can't find the key '" + key + "' in source path!"
+            SgtnClient.logger.error "[Translation][getString]can't find '" + key + "' in '" + component + "'!"
```
[the remark this answers is not in the document]
Comment (linr211) [reply lib/sgtn-client/api/translation.rb]: fixed.

File: lib/sgtn-client/api/source.rb [hunk 8]
```diff
@@ -25,21 +29,34 @@ def self.getSource(component, key, locale)
         end
       end
 
-      def self.getSources(component, locale)
+      def self.getSources(component, locale=SgtnClient::Config.configurations.default)
         SgtnClient.logger.debug "[Source][getSources]component=#{component}, locale=#{locale}"
+        if locale.nil?
+          locale=SgtnClient::Config.configurations.default
+        end
```
[the remark this answers is not in the document]
Comment (linr211) [reply lib/sgtn-client/api/source.rb:36]: As the APIs can be called separately, so has to check it in each APIs.
what's your suggestion?

File: lib/sgtn-client/api/translation.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (linr211) [reply lib/sgtn-client/api/translation.rb]: fixed it by only add the check to get_cs method.

**Comment by linr211** [comment]
> Please spilt this big PR into smaller ones so as to review easily. I don't think all the issues have same priority.


> Please spilt this big PR into smaller ones so as to review easily. I don't think all thea issues have same priority.

the change is not big really, just include some small changes and a async load.

File: lib/sgtn-client/api/translation.rb [hunk 9]
```diff
@@ -86,22 +86,30 @@ def self.getTranslation(component, key, locale)
...
         flocale = SgtnClient::LocaleUtil.fallback(locale)
         cache_key = SgtnClient::CacheUtil.get_cachekey(component, flocale)
         SgtnClient.logger.debug "[Translation][get_cs]cache_key=#{cache_key}"
         expired, items = SgtnClient::CacheUtil.get_cache(cache_key)
-        if items.nil? || expired
-          items = load(component, flocale)
-          if items.nil?
-            items = SgtnClient::Source.getSources(component, SgtnClient::Config.configurations.default)
-            SgtnClient::Core::Cache.put(cache_key, items, 60)
+        t = Thread.new {
```
Comment (jessiejuachon) [inline lib/sgtn-client/api/translation.rb:97]: Isn't this part supposed to be in a different PR?

**Review by jessiejuachon — APPROVED** [review]

File: lib/sgtn-client/api/translation.rb [hunk 10]
```diff
@@ -86,22 +86,30 @@ def self.getTranslation(component, key, locale)
...
+            items = load(component, flocale)
+            if items.nil?
+              items = SgtnClient::Source.getSources(component, SgtnClient::Config.configurations.default)
+              SgtnClient::Core::Cache.put(cache_key, items, 60)
+            else
+              SgtnClient::CacheUtil.write_cache(cache_key, items)
+            end
           else
-            SgtnClient::CacheUtil.write_cache(cache_key, items)
+            SgtnClient.logger.debug "[Translation]get translations from cache with key: " + cache_key
```
Comment (jessiejuachon) [inline lib/sgtn-client/api/translation.rb:107]: Isn't this supposed to be in a different PR?

**Comment by jessiejuachon** [comment]
> > Please spilt this big PR into smaller ones so as to review easily. I don't think all the issues have same priority.
> 
> > Please spilt this big PR into smaller ones so as to review easily. I don't think all thea issues have same priority.
> 
> the change is not big really, just include some small changes and a async load.

**Comment by jessiejuachon** [comment]
Separate PRs for separate concerns please, so can approve smaller parts faster.

**Comment by linr211** [comment]
> Separate PRs for separate concerns please, so can approve smaller parts faster.

which changes should be separated?

File: lib/sgtn-client/api/translation.rb (hunk 9 above)
Comment (linr211) [reply lib/sgtn-client/api/translation.rb:97]: the code logic isn't changed, i think it should be easily understood.

File: lib/sgtn-client/api/translation.rb (hunk 10 above)
Comment (linr211) [reply lib/sgtn-client/api/translation.rb:107]: the logic is not change.

File: lib/sgtn-client/api/translation.rb (hunk 9 above)
Comment (jessiejuachon) [reply lib/sgtn-client/api/translation.rb:97]: Yes, but separate testing needs to happen for this. If this causes an issue, we don;t want the rest of the CloudHealth PR to be pulled due top this,

**Comment by jessiejuachon** [comment]
> 

I had put a comment on every code part that I think is unrelated to the issue.

**Comment by jessiejuachon** [comment]
> > Separate PRs for separate concerns please, so can approve smaller parts faster.
> 
> which changes should be separated?

I had put a comment on every code part that I think is unrelated to the issue.

File: lib/sgtn-client/api/translation.rb (hunk 10 above)
Comment (jessiejuachon) [reply lib/sgtn-client/api/translation.rb:107]: Sure, this is a minor change so not a big deal. It relates to this issue though (@lyiyu66 ): #https://github.com/vmware/singleton/issues/1484
